File: Day-4/labs/java-kafka-consumer-lab/src/main/java/com/kafka/consumer/lab/BasicConsumer.java
package com.kafka.consumer.lab;

import java.time.Duration;
import java.util.Collections;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.clients.consumer.ConsumerRecord;

/**
 * Lab 01 - basic poll loop with manual commit on shutdown.
 */
public final class BasicConsumer {

    public static void main(String[] args) {
        String groupId = ConsumerConfigFactory.arg(args, 0, "lab1-group");
        var props = ConsumerConfigFactory.baseProps(ConsumerConfigFactory.DEFAULT_BOOTSTRAP, groupId);

        try (KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props)) {
            consumer.subscribe(
                    Collections.singletonList(ConsumerConfigFactory.DEMO_TOPIC),
                    new AssignmentLogger("BasicConsumer"));

            Runtime.getRuntime().addShutdownHook(new Thread(() -> {
                System.out.println("Shutdown: committing offsets...");
                consumer.commitSync();
            }));

            System.out.printf("Consuming topic=%s group=%s (Ctrl+C to stop)%n",
                    ConsumerConfigFactory.DEMO_TOPIC, groupId);

            while (true) {
                ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(500));
                for (ConsumerRecord<String, String> record : records) {
                    System.out.printf(
                            "Partition=%d, Offset=%d, Key=%s, Value=%s%n",
                            record.partition(),
                            record.offset(),
                            record.key(),
                            record.value());
                }
            }
        }
    }
}

File: Day-4/labs/python-kafka-consumer-lab/basic_consumer.py
"""Lab 01 - basic Kafka consumer."""
import sys
from kafka import KafkaConsumer
from config import BOOTSTRAP_SERVERS, DEMO_TOPIC


def main():
    group_id = sys.argv[1] if len(sys.argv) > 1 else "lab1-group"
    consumer = KafkaConsumer(
        DEMO_TOPIC,
        bootstrap_servers=BOOTSTRAP_SERVERS,
        group_id=group_id,
        auto_offset_reset="earliest",
        enable_auto_commit=False,
        key_deserializer=lambda k: k.decode("utf-8") if k else None,
        value_deserializer=lambda v: v.decode("utf-8"),
    )
    print(f"Consuming {DEMO_TOPIC} group={group_id} (Ctrl+C to stop)")
    try:
        for msg in consumer:
            print(
                f"Partition={msg.partition}, Offset={msg.offset}, "
                f"Key={msg.key}, Value={msg.value}"
            )
            consumer.commit()
    except KeyboardInterrupt:
        consumer.commit()
    finally:
        consumer.close()


if __name__ == "__main__":
    main()
